### PR TITLE
UCT/COMPONENT: Define and document function types for component methods

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/uct/base/uct_component.h [hunk 1]
```diff
@@ -14,33 +14,127 @@
...
-    ucs_list_link_t        tl_list;            /**< List of uct_md_registered_tl_t */
-    ucs_list_link_t        list;
+/**
+ * Defines a UCT component
+ */
+struct uct_md_component {
+    const char                              name[UCT_MD_COMPONENT_NAME_MAX]; /**< Component name */
+    uct_component_query_md_resources_func_t query_md_resources; /**< Query memory domain resources method */
+    uct_component_md_open_func_t            md_open;            /**< Memory domain open method */
+    uct_component_cm_open_func_t            cm_open;            /**< Connection manager open method */
```
Comment (evgeny-leksikov) [inline src/uct/base/uct_component.h]: why cm_open is a method of "md_component"? shouldn't it be a part of "component"?

Comment (yosefe) [reply src/uct/base/uct_component.h]: i will rename uct_md_component  to uct_component in next PR (was split because of size)

Comment (evgeny-leksikov) [reply src/uct/base/uct_component.h]: Ok, just to be aware, will it be a single structure or some inheritance?

Comment (yosefe) [reply src/uct/base/uct_component.h]: same single structure
just renaming it

Comment (evgeny-leksikov) [reply src/uct/base/uct_component.h]: thanks, as discussed offline it will be a single structure.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11634/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 29 workers (click for details)</summary>

**Note:** the logs will be deleted after 07-Aug-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/6d28d1dff0c95982c06bc9290e5da67f9b85934f/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/c37082ddda92c078a208aab072fe8422babb9f37/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/3e6f68bf439c4126812042c2c9ee452e434fe137/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/9fc696181f571ef0d2a29cdcc103d60050f444e7/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/f98b1afdc3ac20380b20a180aa9431381c1041d5/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/eb37c746fb18bfabde4a44e49e0c9262e2de9f3b/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/45c8b04d0cc65023e2f1623277cad27e5ad01245/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/f1e4eeac603bc4bd6308037c872bac121fed64d5/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/286625f795564a9bba64d4f7a8dcd0c7fab29934/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/9b0da95ee56a807609ba39053b7c2bf9d2b341f3/hpc-test-althca_W0) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/f47cb92cb6d9b98f7040e9f9f221075ad28af2ec/hpc-test-althca_W1) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/04c3a545ff0d81a58cf4ab3ea2b37022dbbadc00/hpc-test-althca_W2) | :heavy_check_mark: SUCCESS
[hpc-test-althca_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/20dc8f85af62664ee12d9b098830e13245f81487/hpc-test-althca_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/a6549769bd2f1e07fe72a9b942ce18c9b36731a6/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/dd02323bf2e1b3012b98b6bc123c158723ac6fec/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/d1e8f3e8881d9c9d5337c596c5eecf71224166f9/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/409f5290861713bd39da6c794225607eda69b957/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/29838ec7f238135ed77769298dfad326ce6c6056/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/7dd4b27838a9c37d16c32b05ec9ef6ec8a0d24ed/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/2a654daf2d0dce24f8d01ca073d8319f14848fba/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/3896793166e78cb819db3ec8bc4bbffa15061107/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/4dbb627f44d87c81a8df17cdd5a038250ee32eb1/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/6df8d9a3e92e7adb61b401eba3e8c1d66dde24a1/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/e1409d5cb40742e7ce406e6d6de6db3db27419cb/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/a70cdbbbac497ecb01bbf59a262883af35534cd4/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/7b3bd2be7a778ac983cf5e08c0f016cf4abde206/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/79d327887f7417fbc7bb5077d2e66ae376a545ef/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/9b558102515b25322fed153c08cd43731fb22665/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/e9d2de97a40e8ced84d2f8fd6f8cc27f/raw/c4e9f4d76e42079ee9f12290391a94d55c965ace/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@evgeny-leksikov can you pls take a look?

**Pull request merged**